test/test_linearizer.py

```diff
@@ -1075,6 +1081,120 @@ def test_local_and_grouped_reduce_multireduce(self):
+          mem_fenced = True
+        if u.uop is UOps.LOAD or u.uop is UOps.STORE and any(v in local_bufs for v in u.vin):
+          assert mem_fenced, "sequential local buffer read/write operations are not fenced"
+          mem_fenced = False
+
+    # parallel reduces
+    a,b = Tensor.rand(64,).realize(), Tensor.rand(64,).realize()
+    r0,r1 = a.sum(), b.sum()
+    ast = _temp_create_multireduce_ast(r0, r1)
+    k = Linearizer(*ast)
+    k.hand_coded_optimizations()
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    k.apply_opt(Opt(OptOps.UNROLL, 0, 2))
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    k.apply_opt(Opt(OptOps.GROUP, 1, 2))
+    check_barriers(k)
+
+    # sequential reduces
+    a,b = Tensor.rand(64,).realize(), Tensor.rand(64,).realize()
+    dummy = Tensor.rand(1,).realize()
+    r0,r1 = (a+dummy).sum(), b.sum()
+    ast = _temp_create_multireduce_ast(r0, r1)
+    k = Linearizer(*ast)
+    k.hand_coded_optimizations()
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    k.apply_opt(Opt(OptOps.UNROLL, 0, 2))
+    check_barriers(k)
+    k = Linearizer(*ast)
+    k.apply_opt(Opt(OptOps.GROUP, 0, 16))
+    k.apply_opt(Opt(OptOps.GROUP, 1, 2))
+    check_barriers(k)
+
+  @unittest.skip("multireduce isn't supported yet")
+  @unittest.skipUnless(Device[Device.DEFAULT].renderer.has_local, "test requires locals")
+  @unittest.skipUnless(Device[Device.DEFAULT].renderer.has_shared, "test requires shared")
+  def test_atomic_store_multireduce(self):
+    # reducops will need to use the local buffer to load the result of a local reduce into every thread, barriers are needed on both sides
+    # of the load to ensure 1) the correct value is in the local buffer and 2) the value isn't overwritten by the next reduceop
+    N = 512
+    Tensor.manual_seed(1882)
+    a,b = Tensor.rand(4,4,N).realize(), Tensor.rand(4,4,N).realize()
+    r0,r1 = a.sum(-1), b.sum(-1)
+    ast = _temp_create_multireduce_ast(r0, r1)
+    helper_linearizer_ast(ast, [a,b], [[Opt(OptOps.GROUP, 0, 2)]])
+
+    # sequential
+    a,b = Tensor.rand(4,4,N).realize(), Tensor.rand(4,4,N).realize()
+    dummy = Tensor.rand(4,4,1).realize()
+    r0,r1 = (a-dummy).sum(-1), b.sum(-1)
+    ast = _temp_create_multireduce_ast(r0, r1, replace_idxs={2:r1}, merge=lambda r0,_: r0)
+    helper_linearizer_ast(ast, [a,b], [[Opt(OptOps.GROUP, 0, 2)]])
+
+  @unittest.skip("multireduce isn't supported yet")
+  @unittest.skipUnless(Device[Device.DEFAULT].renderer.has_local, "test requires locals")
+  @unittest.skipUnless(Device[Device.DEFAULT].renderer.has_shared, "test requires shared")
+  def test_atomic_store_unrolled_multireduce(self):
```

Comment:
this tests an issue with kernels that only have the thread-wise reduce like below. We want to make sure that the first reduce finishes before the data gets loaded and stored in the local buffer for the next reduce (even if they don't depend on each other)

```
  float2 val0 = (float2)(*((__global float2*)(data1+alu1)));
  // float2 val3 = (float2)(*((__global float2*)(data2+alu1))); <--- data for the next reduce
  temp[lidx0] = (val0.y+val0.x);
  // temp[lidx0] = (val3.y+val3.x); <--- we want to make sure this store happens after the first reduce
  barrier(CLK_LOCAL_MEM_FENCE);
  if (alu0) {
    float acc0 = 0.0f;
    for (int ridx0 = 0; ridx0 < 2; ridx0++) {
      float val1 = temp[ridx0];
      acc0 = (val1+acc0);
    }
    temp[0] = acc0;
  }
  barrier(CLK_LOCAL_MEM_FENCE);
  float val2 = temp[0];
```

Reply:
oh yeah another test failing in the `linearizer` branch. Thanks for this!